### PR TITLE
Gererations methods

### PR DESCRIPTION
**Pull request by markhgbrewster**
I have added 4 methods to enable  grouping and return of groups of nodes by their level in the tree structure.  Nodes at the same level in the tree structure would be considered a generation Previously needed similar functionality when us acts_as_tree as part of a permission functionality I was extending. I am not sure if this something that is within the scope of the gem but I enjoyed doing it and I would be interested to hear what you think about it.

### CODE REVIEW

File: lib/acts_as_tree.rb [hunk 1]
```diff
@@ -125,6 +125,13 @@ def self.roots
         end
       EOV
 
+      # Returns a hash of all nodes grouped by their level in the tree structure.
+      #
+      # Class.generations { 0=> [root1, root2], 1=> [root1child1, root1child2, root2child1, root2child2] }
+      def self.generations
+        all.group_by{ |node| node.ancestors.size }
```
Comment (felixbuenemann) [inline lib/acts_as_tree.rb]: This can be simplified to `all.group_by(&:level)`.

File: lib/acts_as_tree.rb [hunk 2]
```diff
@@ -281,6 +288,29 @@ def self_and_siblings
...
+    #  root1child1.generation [root1child2, root2child1, root2child2]
+    def generation
+      self_and_generation - [self]
+    end
+
+    # Returns a reference to the current node and all the nodes at the same level as it in the tree.
+    #
+    #  root1child1.generation [root1child1, root1child2, root2child1, root2child2]
+    def self_and_generation
+      self.class.select {|node| node.ancestors.size == self.ancestors.size }
```
Comment (felixbuenemann) [inline lib/acts_as_tree.rb]: Change the block to `node.level == level`.

**Comment by markhgbrewster** [comment]
@felixbuenemann thanks for the feedback. I have made the changes requested. let me know what you think.

File: lib/acts_as_tree.rb [hunk 3]
```diff
@@ -281,6 +288,27 @@ def self_and_siblings
...
+    end
+
+    # Returns a reference to the current node and all the nodes at the same level as it in the tree.
+    #
+    #  root1child1.generation [root1child1, root1child2, root2child1, root2child2]
+    def self_and_generation
+      self.class.select {|node| node.level == self.level }
+    end
+
+    # Returns level (key) for the current nods generation in the generations hash
```
Comment (felixbuenemann) [inline lib/acts_as_tree.rb]: I think this comment needs to be updated, for example:

> Returns the level (depth) of the current node

I don't think the fact that the level is also a valid key in the generations hash matters in this context.

**Comment by markhgbrewster** [comment]
Thanks again. I have updated the comment as requested.

**Comment by felixbuenemann** [comment]
@markhgbrewster Thanks for your contribution. I've released acts_as_tree 2.6.0 with your changes.

**Comment by markhgbrewster** [comment]
Thanks @felixbuenemann

**Pull request merged**